File: 2099-find-subsequence-of-length-k-with-the-largest-sum/2099-find-subsequence-of-length-k-with-the-largest-sum.cpp
class Solution {
public:
    vector<int> maxSubsequence(vector<int>& nums, int k) {
        int n = nums.size();
        vector<pair<int, int>> vals;        // Auxiliary array

        for (int i = 0; i < n; ++i) vals.push_back({i, nums[i]});

        // Sort by numerical value in descending order
        sort(vals.begin(), vals.end(), [](pair<int, int>& a, pair<int, int>& b) {return a.second > b.second;});
        // Select the first k elements and sort them in ascending order by index
        sort(vals.begin(), vals.begin() + k);
        vector<int> ans;        // target subsequence
        for (int i = 0; i < k; ++i) ans.push_back(vals[i].second);

        return ans;
    }
};

File: 2099-find-subsequence-of-length-k-with-the-largest-sum/2099-find-subsequence-of-length-k-with-the-largest-sum.py
class Solution:
    def maxSubsequence(self, nums: List[int], k: int) -> List[int]:
        ans, min_heap = [], []
        for i, num in enumerate(nums):
            if len(min_heap) >= k:
                if min_heap[0][0] < num:
                    heappop(min_heap)
                    heappush(min_heap, (num, i))
            else:
                heappush(min_heap, (num, i))
        
        min_heap.sort(key = lambda x : x[1])
        for (num, i) in min_heap:
            ans.append(num)

        return ans